### ISSUE TITLE
Installation issue: amrex@23.02+hdf5

### ISSUE THREAD

**Issue by meyersbs** [LLNL/spack#35546]
### Steps to reproduce the issue

```console
$ spack spec -I amrex@23.03%gcc@9.3.0+hdf5
Input spec
--------------------------------
amrex@23.02%gcc@9.3.0+hdf5

Concretized
--------------------------------
ccifr6k  amrex@23.02%gcc@9.3.0~amrdata~cuda~eb~fortran+hdf5~hypre~ipo+linear_solvers+mpi~openmp~particles~petsc~pic~plotfile_tools~rocm~shared~sundials~sycl~tiny_profile build_system=cmake build_type=RelWithDebInfo dimensions=3 precision=double arch=linu
x-rhel7-skylake_avx512
nbe2xqn      ^cmake@3.25.2%gcc@9.3.0~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=linux-rhel7-skylake_avx512
ex5yyfs          ^ncurses@6.4%gcc@9.3.0~symlinks+termlib abi=none build_system=autotools arch=linux-rhel7-skylake_avx512
3g6w4fu          ^openssl@1.1.1t%gcc@9.3.0~docs~shared build_system=generic certs=mozilla arch=linux-rhel7-skylake_avx512
au3vaca              ^ca-certificates-mozilla@2023-01-10%gcc@9.3.0 build_system=generic arch=linux-rhel7-skylake_avx512
vzu47cs              ^perl@5.36.0%gcc@9.3.0+cpanm+open+shared+threads build_system=generic arch=linux-rhel7-skylake_avx512
fsrsn6r                  ^berkeley-db@18.1.40%gcc@9.3.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-rhel7-skylake_avx512
4cm2qmx                  ^bzip2@1.0.8%gcc@9.3.0~debug~pic+shared build_system=generic arch=linux-rhel7-skylake_avx512
fok6ki4                  ^gdbm@1.23%gcc@9.3.0 build_system=autotools arch=linux-rhel7-skylake_avx512
rcybyat                      ^readline@8.2%gcc@9.3.0 build_system=autotools patches=bbf97f1 arch=linux-rhel7-skylake_avx512
lhy5jq7      ^hdf5@1.14.0%gcc@9.3.0~cxx~fortran+hl~ipo~java+mpi+shared~szip~threadsafe+tools api=default build_system=cmake build_type=RelWithDebInfo patches=0b5dd6f arch=linux-rhel7-skylake_avx512
xp45tof          ^pkgconf@1.8.0%gcc@9.3.0 build_system=autotools arch=linux-rhel7-skylake_avx512
pv3fue4          ^zlib@1.2.13%gcc@9.3.0+optimize+pic+shared build_system=makefile arch=linux-rhel7-skylake_avx512
v6i343e      ^mvapich@3.0a%gcc@9.3.0~alloca~cuda~debug+regcache+wrapperrpath build_system=autotools ch3_rank_bits=32 file_systems=auto netmod=ofi process_managers=auto threads=multiple arch=linux-rhel7-skylake_avx512
bi4bnqn          ^bison@3.8.2%gcc@9.3.0 build_system=autotools arch=linux-rhel7-skylake_avx512
oud47uw              ^diffutils@3.8%gcc@9.3.0 build_system=autotools arch=linux-rhel7-skylake_avx512
a2xcvvp              ^m4@1.4.19%gcc@9.3.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-rhel7-skylake_avx512
rpgflvg                  ^libsigsegv@2.13%gcc@9.3.0 build_system=autotools arch=linux-rhel7-skylake_avx512
ymspxqm          ^findutils@4.9.0%gcc@9.3.0 build_system=autotools patches=440b954 arch=linux-rhel7-skylake_avx512
uzxnsoe          ^libfabric@1.16.1%gcc@9.3.0~debug~kdreg build_system=autotools fabrics=sockets,tcp,udp arch=linux-rhel7-skylake_avx512
jks3j5x          ^libpciaccess@0.16%gcc@9.3.0 build_system=autotools arch=linux-rhel7-skylake_avx512
qxy2bgr              ^libtool@2.4.7%gcc@9.3.0 build_system=autotools arch=linux-rhel7-skylake_avx512
uisde6u              ^util-macros@1.19.3%gcc@9.3.0 build_system=autotools arch=linux-rhel7-skylake_avx512
u5yxcrw          ^libxml2@2.10.3%gcc@9.3.0~python build_system=autotools arch=linux-rhel7-skylake_avx512
boo6zb3              ^libiconv@1.17%gcc@9.3.0 build_system=autotools libs=shared,static arch=linux-rhel7-skylake_avx512
dc4igz5              ^xz@5.4.1%gcc@9.3.0~pic build_system=autotools libs=shared,static arch=linux-rhel7-skylake_avx512
```

### Error message

<details><summary>Error message</summary><pre>
==> Installing amrex-23.02-ccifr6kx7zxeqv3k7mif25o6gf7xkcw5
==> No binary for amrex-23.02-ccifr6kx7zxeqv3k7mif25o6gf7xkcw5 found: installing from source
==> Using cached archive: /dev/shm/bsmits_spack/var/spack/cache/_source-cache/archive/f4/f443c5eb4b89f4a74bf0e1b8a5943da18ab81cdc76aff12e8282ca43ffd06412.tar.gz
==> No patches needed for amrex
==> amrex: Executing phase: 'cmake'
==> Error: ProcessError: Command exited with status 1:
    '/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/cmake-3.25.2-nbe2xqn25boaqufnsaeqtdnil4pfwpgy/bin/cmake' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:STRING=/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/a
mrex-23.02-ccifr6kx7zxeqv3k7mif25o6gf7xkcw5' '-DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo' '-DBUILD_TESTING:BOOL=OFF' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=ON' '-DCMAKE_I
NSTALL_RPATH:STRING=/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/amrex-23.02-ccifr6kx7zxeqv3k7mif25o6gf7xkcw5/lib;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/amrex-23.02-ccifr6kx7zxeqv3k7mif25o6gf7xkcw5/li
b64;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/hdf5-1.14.0-lhy5jq7rjm6psfjbs7vde5ny57yhour5/lib;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/mvapich-3.0a-v6i343egxor3ggvm5ifk53vhlzvtcunx/lib;/dev/shm/bsmi
ts_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/libfabric-1.16.1-uzxnsoe27thisgcp3yo3py3ffgewsxju/lib;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/libpciaccess-0.16-jks3j5xtmwd4jof5nbotydlyvabd55xc/lib;/dev/shm/bsmits_spac
k/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/libxml2-2.10.3-u5yxcrw5qcck3o5u2dtlqkwrjmezeorg/lib;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/libiconv-1.17-boo6zb3seinlrqpvirr44waxik4atl3b/lib;/dev/shm/bsmits_spack/opt/spack/l
inux-rhel7-skylake_avx512/gcc-9.3.0/xz-5.4.1-dc4igz55iuvtzjuckgqr2eedllfrkjxe/lib;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/zlib-1.2.13-pv3fue4yfrg2wwftonn6mhox6oworevh/lib' '-DCMAKE_PREFIX_PATH:STRING=/dev/shm/bsmits_spack/opt
/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/hdf5-1.14.0-lhy5jq7rjm6psfjbs7vde5ny57yhour5;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/mvapich-3.0a-v6i343egxor3ggvm5ifk53vhlzvtcunx;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylak
e_avx512/gcc-9.3.0/libxml2-2.10.3-u5yxcrw5qcck3o5u2dtlqkwrjmezeorg;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/xz-5.4.1-dc4igz55iuvtzjuckgqr2eedllfrkjxe;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/libpcia
ccess-0.16-jks3j5xtmwd4jof5nbotydlyvabd55xc;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/libfabric-1.16.1-uzxnsoe27thisgcp3yo3py3ffgewsxju;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/cmake-3.25.2-nbe2xqn25
boaqufnsaeqtdnil4pfwpgy;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/zlib-1.2.13-pv3fue4yfrg2wwftonn6mhox6oworevh;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/libiconv-1.17-boo6zb3seinlrqpvirr44waxik4atl3b'
 '-DUSE_XSDK_DEFAULTS=ON' '-DAMReX_SPACEDIM:STRING=3' '-DBUILD_SHARED_LIBS:BOOL=OFF' '-DAMReX_MPI:BOOL=ON' '-DAMReX_OMP:BOOL=OFF' '-DXSDK_PRECISION:STRING=DOUBLE' '-DXSDK_ENABLE_Fortran:BOOL=OFF' '-DAMReX_FORTRAN_INTERFACES:BOOL=OFF' '-DAMReX_EB:BOOL=OFF
' '-DAMReX_LINEAR_SOLVERS:BOOL=ON' '-DAMReX_AMRDATA:BOOL=OFF' '-DAMReX_PARTICLES:BOOL=OFF' '-DAMReX_PLOTFILE_TOOLS:BOOL=OFF' '-DAMReX_TINY_PROFILE:BOOL=OFF' '-DAMReX_HDF5:BOOL=ON' '-DAMReX_HYPRE:BOOL=OFF' '-DAMReX_PETSC:BOOL=OFF' '-DAMReX_SUNDIALS:BOOL=O
FF' '-DAMReX_PIC:BOOL=OFF' '/dev/shm/spack/stage/bsmits/spack-stage-amrex-23.02-ccifr6kx7zxeqv3k7mif25o6gf7xkcw5/spack-src'

1 error found in build log:
     31    -- Check if compiler accepts -pthread - yes
     32    -- Found Threads: TRUE
     33    -- Found MPI_C: /dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/mvapich-3.0a-v6i343egxor3ggvm5ifk53vhlzvtcunx/lib/libmpi.so (found version "3.1")
     34    -- Found MPI_CXX: /dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/mvapich-3.0a-v6i343egxor3ggvm5ifk53vhlzvtcunx/lib/libmpicxx.so (found version "3.1")
     35    -- Found MPI: TRUE (found version "3.1") found components: C CXX
     36    -- Found HDF5: /dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/hdf5-1.14.0-lhy5jq7rjm6psfjbs7vde5ny57yhour5/lib/libhdf5.so;/dev/shm/bsmits_spack/opt/spack/linux-rhel7-skylake_avx512/gcc-9.3.0/mvapich-3.0a-v6i343egxor3ggvm
           5ifk53vhlzvtcunx/lib/libmpi.so (found suitable version "1.14.0", minimum required is "1.10.4")
  >> 37    CMake Error at Tools/CMake/AMReXThirdPartyLibraries.cmake:10 (message):
     38    
     39    
     40      HDF5 library does not support parallel I/O
     41    Call Stack (most recent call first):
     42      Src/CMakeLists.txt:28 (include)
     43    

See build log for details:
  /dev/shm/spack/stage/bsmits/spack-stage-amrex-23.02-ccifr6kx7zxeqv3k7mif25o6gf7xkcw5/spack-build-out.txt
</pre></details>

### Information on your system

```
$ spack debug report
* **Spack:** 0.20.0.dev0
* **Python:** 3.6.8
* **Platform:** linux-rhel7-skylake_avx512
* **Concretizer:** clingo
```

### Additional information

- [amrex-CMakeError.log](https://github.com/spack/spack/files/10768881/amrex-CMakeError.log)
- [amrex-CMakeOutput.log](https://github.com/spack/spack/files/10768882/amrex-CMakeOutput.log)
- [amrex-spack-build-env.txt](https://github.com/spack/spack/files/10768883/amrex-spack-build-env.txt)
- [amrex-spack-build-out.txt](https://github.com/spack/spack/files/10768884/amrex-spack-build-out.txt)

@WeiqunZhang 
@asalmgren 
@etpalmer63 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by aweits** [comment]
If I change cmake's FindHDF5 modulefile to something like this I can resolve the issue:

```diff
[aweits@coltrane cmake]$ git diff
diff --git a/Modules/FindHDF5.cmake b/Modules/FindHDF5.cmake
index db03c54e85e4..e9caa91d3202 100644
--- a/Modules/FindHDF5.cmake
+++ b/Modules/FindHDF5.cmake
@@ -511,7 +511,7 @@ if(NOT HDF5_FOUND AND NOT HDF5_NO_FIND_PACKAGE_CONFIG_FILE)
         if(HDF5_FIND_DEBUG)
             message(STATUS "Found HDF5 at ${HDF5_DIR} via NO_MODULE. Now trying to extract locations etc.")
         endif()
-        set(HDF5_IS_PARALLEL ${HDF5_ENABLE_PARALLEL})
+        set(HDF5_IS_PARALLEL ${HDF5_ENABLE_PARALLEL} CACHE BOOL "HDF5 library is compiled with parallel IO support")
         set(HDF5_INCLUDE_DIRS ${HDF5_INCLUDE_DIR})
         set(HDF5_LIBRARIES)
         if (NOT TARGET hdf5 AND NOT TARGET hdf5-static AND NOT TARGET hdf5-shared)
```

**Comment by WeiqunZhang** [comment]
If HDF5 is built without shared library support, it seems to work.

**Comment by WeiqunZhang** [comment]
Looks like a bug in FindHDF5.  Here is a merge request.  https://gitlab.kitware.com/cmake/cmake/-/merge_requests/8234

**Comment by WeiqunZhang** [comment]
The issue has been resolved in amrex@develop and amrex@23.03 (#35765).  Furthermore, the merge request to CMake has been merged, and the fix will appear in CMake 3.27.0.  So we can probably close this issue now.
